File: backend/Timesheets.Api/Timesheets/Endpoints/AllocateTimesheet.cs
using Microsoft.AspNetCore.Http.HttpResults;
using Microsoft.AspNetCore.Mvc;
using Timesheets.Api.Auth;
using Timesheets.Api.Common.Extensions;
using Timesheets.Api.Data;
using Timesheets.Api.Employees;
using Timesheets.Api.Timesheets.Allocation;

namespace Timesheets.Api.Timesheets.Endpoints;

public sealed class AllocateTimesheet : IEndpoint
{
    public static void Map(IEndpointRouteBuilder app) =>
        app.MapPost("/{id}/allocate", Handle)
            .WithSummary("Allocate Timesheet Edit")
            .WithRequestValidation<TimesheetEditRequest>();

    public sealed record ProjectCell(decimal Hours, bool Locked);
    public sealed record DayResponse(DateTime Date, int?[] Work, int?[] Break, decimal CoreHours, IReadOnlyDictionary<Guid, ProjectCell> ProjectCells, bool AttendanceAdjusted);
    public sealed record Response(IReadOnlyList<DayResponse> Days, TimesheetEvaluation Evaluation);

    private static async Task<Results<Ok<Response>, NotFound, ForbidHttpResult>> Handle(Guid id, [FromQuery] int? day, [FromBody] TimesheetEditRequest request, AppDbContext dbContext, ICurrentUser user, CancellationToken cancellationToken)
    {
        LoadedTimesheet? loaded = await TimesheetEngine.LoadAsync(id, dbContext, cancellationToken);
        if (loaded is null)
        {
            return TypedResults.NotFound();
        }
        if ((!user.IsGlobalManagerRole() && user.EmployeeId != loaded.Timesheet.EmployeeId) || loaded.Timesheet.TimesheetStatusId != TimesheetWorkflow.DraftStatusId)
        {
            return TypedResults.Forbid();
        }

        EditableTimesheet sheet = TimesheetEngine.BuildEditableTimesheet(loaded, request);

        if (EmployeeTypes.TracksAttendance(loaded.EmployeeTypeId))
        {
            NonAcademicTimesheetAllocator allocator = new(loaded, sheet);
            if (day is int dayNumber)
            {
                allocator.AllocateDay(dayNumber);
            }
            else
            {
                allocator.AllocateMonth();
            }
        }
        else
        {
            AcademicTimesheetAllocator allocator = new(loaded, sheet);
            if (day is int dayNumber)
            {
                allocator.AllocateDay(dayNumber);
            }
            else
            {
                allocator.AllocateMonth();
            }
        }

        return TypedResults.Ok(CreateAllocationResponse(loaded, sheet));
    }

    private static Response CreateAllocationResponse(LoadedTimesheet loaded, EditableTimesheet sheet)
    {
        List<DayResponse> allocation = sheet.Days
            .Select(day => new DayResponse(
                Date: day.Date,
                Work: [ConvertToMinutes(day.ClockIn), ConvertToMinutes(day.ClockOut)],
                Break: [ConvertToMinutes(day.BreakStart), ConvertToMinutes(day.BreakEnd)],
                CoreHours: day.CoreHours,
                ProjectCells: sheet.Projects.ToDictionary(
                    project => project.Id,
                    project => new ProjectCell(
                        day.ProjectHours.GetValueOrDefault(project.Id),
                        day.ProjectHoursFixed.GetValueOrDefault(project.Id))),
                AttendanceAdjusted: day.AttendanceAdjusted))
            .ToList();
        return new Response(Days: allocation, Evaluation: TimesheetEngine.Evaluate(loaded, sheet));
    }

    private static int? ConvertToMinutes(TimeSpan? value) => value.HasValue ? (int)Math.Round(value.Value.TotalMinutes) : null;
}

File: frontend/src/features/timesheet/timesheet/api/allocateTimesheet.ts
import { ApiUrl, customFetch } from "@/constants/api";
import type { Timesheet, TimesheetEvaluation } from "../../Timesheet";
import { type ApiTimesheetEvaluation, buildTimesheetDraft, mapTimesheetEvaluation } from "./timesheetDraft";

interface ApiAllocationDay {
  work: [number | null, number | null];
  break: [number | null, number | null];
  coreHours: number;
  projectHours: Record<string, number>;
}

interface ApiAllocation {
  days: ApiAllocationDay[];
  evaluation: ApiTimesheetEvaluation;
}

export interface TimesheetAllocation {
  days: AllocationDay[];
  evaluation: TimesheetEvaluation;
}

interface AllocationDay {
  clockIn: string;
  clockOut: string;
  breakStart: string;
  breakEnd: string;
  coreHours: number;
  projectHours: Record<string, number>;
}

const pad2 = (value: number) => value.toString().padStart(2, "0");
const minutesToTime = (value: number | null | undefined) => {
  if (value == null || value < 0) return "";
  const hours = Math.floor(value / 60);
  const minutes = value % 60;
  return `${pad2(hours)}:${pad2(minutes)}`;
};

export const allocateTimesheet = async (timesheet: Timesheet, day?: number): Promise<TimesheetAllocation> => {
  const query = day ? `?day=${day}` : "";
  const allocation = await customFetch<ApiAllocation>(`${ApiUrl}/timesheets/${timesheet.id}/allocate${query}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(buildTimesheetDraft(timesheet)),
  });
  return {
    days: allocation.days.map((day) => ({
      clockIn: minutesToTime(day.work?.[0]),
      clockOut: minutesToTime(day.work?.[1]),
      breakStart: minutesToTime(day.break?.[0]),
      breakEnd: minutesToTime(day.break?.[1]),
      coreHours: day.coreHours,
      projectHours: day.projectHours,
    })),
    evaluation: mapTimesheetEvaluation(allocation.evaluation),
  };
};
